File: jvm/src/leetcode/GenerateParentheses.java
package leetcode;

import java.util.ArrayList;
import java.util.List;

public class GenerateParentheses {

    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        backtrack(ans, new StringBuilder(), 0, 0, n);
        return ans;
    }


    private void backtrack(List<String> set, StringBuilder sb, int op, int cp, int n) {
        // base case
        if (sb.length() == n * 2) {
            set.add(sb.toString());
        } else {
            if (op < n) {
                backtrack(set, sb.append("("), op + 1, cp, n);
                sb.deleteCharAt(sb.length() - 1);
            }

            if (op > cp) {
                backtrack(set, sb.append(")"), op + 1, cp, n);
                sb.deleteCharAt(sb.length() - 1);
            }
        }
    }
}

File: scala/src/main/scala/leetcode/GenerateParentheses.scala
package leetcode

object GenerateParentheses {
  /*
        f(0) - ""
        f(1) - (f(0)) - "()"
        f(2) - (f(0)) + f(1), (f(1))
        f(3) - ... (f(1)) + f(1)

        f(n) - ( f(k) ) + f(n-k-1), 0 <= k < n
    */
  def generateParenthesis(n: Int): List[String] =
      if (n == 0) List("")
      else
        for {
          k <- (0 until n).toList
          a <- generateParenthesis(k)
          b <- generateParenthesis(n - k - 1)
        } yield "(" + a + ")" + b
}
